docs: splits OIDC and JWT auth method documentation to distinguish workflows (#22451)
hashicorp/vault: website/content/docs/auth/jwt.mdx

---
layout: docs
page_title: JWT - Auth Method
description: >-
  The JWT auth method allows authentication using signed JWTs
---

# JWT auth method

@include 'x509-sha1-deprecation.mdx'

> The `jwt` auth method can be used to authenticate with Vault using JSON Web Tokens
> ([JWT](https://en.wikipedia.org/wiki/JSON_Web_Token)). This form of authentication
> is commonly used in Workload Identity Federation (WIF) workflows for machine-to-machine
> authentication.

Reviewer: ```suggestion
Use the `jwt` authN method to authenticate to Vault with JSON Web Tokens
([JWT](https://en.wikipedia.org/wiki/JSON_Web_Token)). JWT authentication
is commonly used in Workload Identity Federation (WIF) workflows for
machine-to-machine authentication.
```

Style correction: avoid "this" as a pronoun, write in active voice with directive statements.

Author: Note that I'm changing "authN" back to "auth method" since this is how they're generally referred to.

> The auth method is configured to trust a JWT issuer. For example, this could be a
> cloud service provider that makes workload identity available in a virtual machine
> or CI provider that provides identity to jobs. It must be [configured](#jwt-verification)
> with public keys from the JWT issuer by either OIDC Discovery, a JSON Web Key Set (JWKS) URL,
> or locally-provided keys. A signed JWT from the configured issuer is required as the form
> of authentication. The auth method will cryptographically verify the signature using the
> issuer's public key. The verified JWT claims can be used to provide [further validation](#bound-claims)
> to gate authentication to Vault.

Reviewer: ```suggestion
To use JWT, you must configure Vault to trust a JWT issuer. For example, a
cloud service provider that makes workload identity available in a virtual machine
or CI provider that provides identities to jobs.

Vault will require a signed JWT from the configured issuer as the form of
authentication and the authN method will cryptographically verify the client
signature using the issuer's public key. You can use verified JWT claims to
gate authentication to Vault by requiring [additional validation](#bound-claims).

You must [configure](#jwt-verification) the authN method with public keys from
the JWT issuer with OIDC Discovery, a JSON Web Key Set (JWKS) URL, or
locally-provided keys. 
```

Style correction: write in active voice, make content scannable, make it clear who/what takes action

## Authentication

> ### Via the CLI

Reviewer: ```suggestion
### Authenticate with the CLI
```

Style correction: Use directive language for headings, make headings friendly to screen readers

```shell-session
$ vault write auth/jwt/login role=demo jwt=...

> ```

Reviewer: How is this example different from the default command listed below?

> The default path for the JWT authentication backend is `/jwt`, so if you're using the
> default backend, the command would be:

Reviewer: ```suggestion
The default path for the JWT authentication backend is `/jwt`. To use the default backend, set the path parameter to `auth/jwt/login` in your `vault write` command:
```

Style correction: write in active voice, 1 idea per sentence.

Note: it wasn't obvious from the example command how/where the default value was being used so my edits may not be technically accurate. It would help if we provided an example of a non-default command as well.

```shell-session
$ vault write auth/jwt/login role=demo jwt=...
```

If your JWT auth backend is using a different path, use that path.

> ### Via the API

Reviewer: ```suggestion
### Authenticate with the API
```

Style correction: Use directive language for headings, make headings friendly to screen readers

```shell-session
$ curl \
    --request POST \
    --data '{"jwt": "your_jwt", "role": "demo"}' \
    http://127.0.0.1:8200/v1/auth/jwt/login
```

The response will contain a token at `auth.client_token`:

```json
{
  "auth": {
    "client_token": "hvs.abcd",
    "accessor": "78e87a38-84ed-2692-538f-ca8b9f400ab3",
    "policies": ["default"],
    "metadata": {
      "role": "demo"
    },
    "lease_duration": 2764800,
    "renewable": true
  }
}
```

## Configuration

Auth methods must be configured in advance before users or machines can
authenticate. These steps are usually completed by an operator or configuration
management tool.

1. Enable the JWT auth method.

   ```text
   $ vault auth enable jwt
   ```

1. Use the `/config` endpoint to configure Vault. To support JWT roles, either local keys,
   a JWKS URL, or an OIDC Discovery URL must be present. For an explanation of available
   configuration options, see the [API documentation](/vault/api-docs/auth/jwt#jwt-verification)
   section.

   ```text
   $ vault write auth/jwt/config \
      oidc_discovery_url="https://MYDOMAIN.eu.auth0.com/"
  ```

1. Create a named role:

   ```text

>    vault write auth/jwt/role/demo \

Reviewer 2: Noticed some inconsistent use of `$`. I am not sure what the recommended format is? With or without `$` in front of commands?

Author: @schavis - Curious if you have a preference for this?

Reviewer: Generally, we should use the `$` in example output to differentiate the command from the output text. But we should leave it off any example where we expect folks to copy/paste the command.
